elyra/pipeline/processor.py

```diff
@@ -148,30 +158,48 @@ def to_json(self):
 
 class PipelineProcessor(LoggingConfigurable):  # ABC
 
-    _type = None
+    _type: str = None
 
-    root_dir = Unicode(allow_none=True)
+    _component_registry: ComponentRegistry = None
 
-    component_registry: ComponentRegistry = ComponentRegistry()
+    root_dir = Unicode(allow_none=True)
 
     enable_pipeline_info = Bool(config=True,
                                 default_value=(os.getenv('ELYRA_ENABLE_PIPELINE_INFO', 'true').lower() == 'true'),
                                 help="""Produces formatted logging of informational messages with durations
                                 (default=True). (ELYRA_ENABLE_PIPELINE_INFO env var)""")
 
     def __init__(self, root_dir, **kwargs):
-        super(PipelineProcessor, self).__init__(**kwargs)
+        super().__init__(**kwargs)
         self.root_dir = root_dir
 
     @property
     @abstractmethod
-    def type(self):
+    def type(self) -> str:
         raise NotImplementedError()
 
-    def get_components(self):
-        components = self.component_registry.get_all_components(processor_type=self.type)
+    def get_components(self) -> List[Component]:
+        """
+        Retrieve components common to all runtimes
+        """
+        components: List[Component] = ComponentRegistry.get_generic_components()
+
+        # Retrieve runtime-specific components
+        if self._component_registry:
+            components.extend(self._component_registry.get_all_components())
+
         return components
 
+    def get_component(self, component_id: str) -> Component:
+        """
+        Retrieve runtime-specific component_id details if component_id is not one of the generic set
+        """
+
+        if component_id not in ('notebooks', 'python-script', 'r-script'):
```

Comment:
Nit: not future proof. Should use key lookup  in `_generic_components` instead of hard-coding them here.

Reply:
Will rework this in a future pr